public/app/features/dashboard/state/PanelQueryState.ts

```diff
@@ -100,60 +124,169 @@ export class PanelQueryState {
             this.sendSeries = true;
           }
 
-          // Make sure the response is in a supported format
-          const series = this.sendSeries ? getProcessedSeriesData(resp.data) : [];
-          const legacy = this.sendLegacy
-            ? resp.data.map(v => {
-                if (isSeriesData(v)) {
-                  return toLegacyResponseData(v);
-                }
-                return v;
-              })
-            : undefined;
-
-          resolve(
-            (this.data = {
-              state: LoadingState.Done,
-              request: this.request,
-              series,
-              legacy,
-            })
-          );
+          // Save the result state
+          this.response = {
+            state: LoadingState.Done,
+            request: this.request,
+            series: this.sendSeries ? getProcessedSeriesData(resp.data) : [],
+            legacy: this.sendLegacy ? translateToLegacyData(resp.data) : undefined,
+          };
+          resolve(this.getPanelData());
         })
         .catch(err => {
           resolve(this.setError(err));
         });
-    }));
+    });
+
+    return this.executor;
+  }
+
+  // Send a notice when the stream has updated the current model
+  streamCallback: () => void;
+
+  // This gets all stream events and keeps track of them
+  // it will then delegate real changes to the PanelQueryRunner
+  dataStreamObserver: DataStreamObserver = (stream: DataStreamState) => {
+    // Streams only work with the 'series' format
+    this.sendSeries = true;
+
+    // Add the stream to our list
+    let found = false;
+    const active = this.streams.map(s => {
```

Comment:
wonder if streams would work better as object map (by key) instead of an array. Maybe makes other stuff clunky, not sure.

Reply:
I tried map... But consistent order is important. And we have to loop on the array anyway.  When you add to theap it should not change the order of existing series

Reply:
ah, makes sense, array it will be then 👍